deckhouse-controller/pkg/controller/deckhouse-release/controller.go

```diff
@@ -0,0 +1,455 @@
+		return ctrl.Result{}, nil
+	}
+
+	return r.createOrUpdateReconcile(ctx)
+}
+
+func (r *deckhouseReleaseReconciler) createOrUpdateReconcile(ctx context.Context) (ctrl.Result, error) {
+	deckhousePods, err := r.getDeckhousePods(ctx)
+	if err != nil {
+		r.logger.Warnf("Error getting deckhouse pods: %s", err)
+		return ctrl.Result{RequeueAfter: defaultCheckInterval}, nil
+	}
+
+	if len(deckhousePods) == 0 {
+		r.logger.Warn("Deckhouse pods not found. Skipping update")
+		return ctrl.Result{RequeueAfter: defaultCheckInterval}, nil
+	}
+
+	discoveryData, err := r.getDeckhouseDiscoveryData(ctx)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("get release channel: %w", err)
+	}
+
+	if len(discoveryData.ReleaseChannel) == 0 {
+		return r.tagUpdate(ctx, deckhousePods)
+	}
+
+	releaseData, err := r.getReleaseData(ctx)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("get release data: %w", err)
+	}
+
+	podReady := r.isDeckhousePodReady()
+	deckhouseUpdater, err := d8updater.NewDeckhouseUpdater(r.logger, r.client, r.dc, discoveryData, r.updatePolicy.Get().Update.Mode, releaseData, podReady)
+
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("initializing deckhouse updater: %w", err)
+	}
+
+	if podReady {
+		if releaseData.IsUpdating {
+			_ = deckhouseUpdater.ChangeUpdatingFlag(false)
+		}
+	}
+
+	var releases v1alpha1.DeckhouseReleaseList
+	err = r.client.List(ctx, &releases)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("get deckhouse releases: %w", err)
+	}
+
+	pointerReleases := make([]*v1alpha1.DeckhouseRelease, 0, len(releases.Items))
+	for _, r := range releases.Items {
+		pointerReleases = append(pointerReleases, &r)
+	}
+	deckhouseUpdater.PrepareReleases(pointerReleases)
+	if deckhouseUpdater.ReleasesCount() == 0 {
+		return ctrl.Result{RequeueAfter: defaultCheckInterval}, nil
+	}
+
+	// predict next patch for Deploy
+	deckhouseUpdater.PredictNextRelease()
+
+	// has already Deployed the latest release
+	if deckhouseUpdater.LastReleaseDeployed() {
+		return ctrl.Result{RequeueAfter: defaultCheckInterval}, nil
+	}
+
+	// some release is forced, burn everything, apply this patch!
+	if deckhouseUpdater.HasForceRelease() {
+		deckhouseUpdater.ApplyForcedRelease()
+		return ctrl.Result{RequeueAfter: defaultCheckInterval}, nil
+	}
+
+	if deckhouseUpdater.PredictedReleaseIsPatch() {
+		// patch release does not respect update windows or ManualMode
+		deckhouseUpdater.ApplyPredictedRelease(nil)
+		return ctrl.Result{RequeueAfter: defaultCheckInterval}, nil
+	}
+
+	var windows update.Windows
+	if !deckhouseUpdater.InManualMode() {
+		windows = r.updatePolicy.Get().Update.Windows
+	}
+
+	deckhouseUpdater.ApplyPredictedRelease(windows)
+	return ctrl.Result{RequeueAfter: defaultCheckInterval}, nil
+}
+
+func (r *deckhouseReleaseReconciler) getDeckhousePods(ctx context.Context) ([]corev1.Pod, error) {
+	var pods corev1.PodList
+	err := r.client.List(
```

Comment:
Guess you have to add them [here](https://github.com/deckhouse/deckhouse/blob/main/deckhouse-controller/pkg/controller/controller.go#L131)

Reply:
Done